snmpproxy/data.go

```diff
@@ -0,0 +1,148 @@
+package snmpproxy
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/soniah/gosnmp"
+)
+
+type RequestType string
+
+func (t *RequestType) UnmarshalJSON(data []byte) error {
+	var s string
+	if err := json.Unmarshal(data, &s); err != nil {
+		return fmt.Errorf("RequestType should be a string, got %s: %w", string(data), err)
+	}
+
+	*t = RequestType(s)
+
+	switch *t {
+	case Get, GetNext, Walk:
+		return nil
+	case "":
+		return errors.New("RequestType mustn't be empty")
+	default:
+		return fmt.Errorf("unknown RequestType \"%s\"", *t)
+	}
+}
+
+const (
+	Get     = RequestType("get")
+	GetNext = RequestType("getNext")
+	Walk    = RequestType("walk")
+)
+
+type SnmpVersion gosnmp.SnmpVersion
+
+func (v *SnmpVersion) UnmarshalJSON(data []byte) error {
+	var s string
+	if err := json.Unmarshal(data, &s); err != nil {
+		return fmt.Errorf("snmpVersion should be a string, got %s: %w", string(data), err)
+	}
+
+	switch s {
+	case "1":
+		*v = SnmpVersion(gosnmp.Version1)
+	case "2c":
+		*v = SnmpVersion(gosnmp.Version2c)
+	case "":
+		return errors.New("snmpVersion mustn't be empty")
+	default:
+		return fmt.Errorf("unknown or unsupported snmpVersion \"%s\", supported are: 1, 2c", s)
+	}
+
+	return nil
+}
+
+type Request struct {
+	RequestType RequestType   `json:"request_type"`
+	Host        string        `json:"host"`
+	Community   string        `json:"community"`
+	Version     SnmpVersion   `json:"version"`
+	Retries     uint8         `json:"retries"`
+	Timeout     time.Duration `json:"timeout"`
+	Oids        []string      `json:"oids"`
+}
+
+func (r *Request) UnmarshalJSON(data []byte) error {
+	type tmp Request
+
+	var t tmp
+
+	if err := json.Unmarshal(data, &t); err != nil {
+		return fmt.Errorf("failed to unmarshal request body into Request struct, got %+v: %w", string(data), err)
+	}
+
+	if t.RequestType == "" {
+		return fmt.Errorf("field request_type mustn't be empty")
+	}
+
+	if t.RequestType == Walk && len(t.Oids) > 1 {
+		return fmt.Errorf("only single OID is supported with RequestType = Walk, got %d", len(t.Oids))
+	}
+
+	if t.Host == "" {
+		return fmt.Errorf("field host mustn't be empty")
+	}
+
+	if t.Community == "" {
+		t.Community = "public"
+	}
+
+	if len(t.Oids) == 0 {
+		return fmt.Errorf("at least one OID must be provided")
+	}
+
+	for _, oid := range t.Oids {
+		if oid[0] != '.' {
+			return fmt.Errorf("all OIDs must begin with a dot")
+		}
+	}
+
+	if t.Version == 0 {
+		var v struct {
+			Version string `json:"version"`
+		}
+
+		_ = json.Unmarshal(data, &v)
+
+		if v.Version == "" {
+			return fmt.Errorf("field version mustn't be empty")
+		}
+	}
+
+	if t.Timeout > 300 {
+		return fmt.Errorf("maximum allowed timeout is 300 seconds, got %d seconds", t.Timeout)
+	}
+
+	if t.Timeout == 0 {
+		t.Timeout = 2
+	}
+
+	t.Timeout *= time.Second
+
+	if t.Retries > 10 {
+		return fmt.Errorf("maximum allowed number of retries is 10, got %d", t.Retries)
```

Comment:
why this?

Reply:
stejně jako timeout - nechceš aby ti někdo hodil neexistující adresu a ty ses jí snažil pingovat do nekonečna ... už 5 retries je dost, víc jak 10 je unreal

Reply:
taky bych klidne povolil